### DISCUSSION TITLE
Support obs-websocket v5.0 protocol?

### DISCUSSION THREAD

**Discussion by ianmjones** [muesli/obs-cli#55] [Ideas]
Looks like obs-websocket v5.0 has hit beta and includes a new version of the protocol.

I tried building both obs-cli v0.5.0 and the latest `master` commit and running them against OBS Studio with obs-websocket v5.0.0-alpha3 (latest available on NixOS unstable), this is what I saw:

```
obs-cli scene list
error: Failed auth: Client/server version mismatch? Unrecognized message: {"d":{"obsWebSocketVersion":"5.0.0","rpcVersion":1},"op":0}
```

Is there a plan to support the new protocol? (I have no idea how big of a question this might be :smile: )

https://github.com/obsproject/obs-websocket/releases/tag/5.0.0-beta1

**Comment by ianmjones** [comment]
Just had a little poke around to see whether this is something I could help with, but, umm, no, this is waaaaay out of my league as I'm a beginner with Go! :grimacing: 

I did notice that the core protocol stuff is actually in a separate library, and there's already [some work underway there](https://github.com/andreykaipov/goobs/tree/feature/v5) to support obs-websocket v5.0, which is great to see. And of course, that too is way out of my league to help with at the moment.

Maybe further down the line I can get involved in some way.

**Comment by muesli** [comment]
@ianmjones There's a protocol definition and a code generator in `goobs` that processes the file. I assume @andreykaipov is already working on upgrading to v5.0 in the near future.

**Comment by andreykaipov** [comment]
Hi y'all - yes I started work for supporting v5 a while ago, but had to drop it because the v5 protocol.json isn't as complete as the v4 one yet. I asked in the project's Discord a while back:

<img src="https://user-images.githubusercontent.com/9457739/162336974-fd8878ed-81ab-4193-9e64-2d9b2fc287ee.png" width="600" />

Links from the image:

https://github.com/obsproject/obs-websocket/blob/4.x-current/docs/generated/protocol.md#getsceneitemlist
https://github.com/obsproject/obs-websocket/blob/master/docs/generated/protocol.md#getsceneitemlist

I think the realization of having to manually create however many structs kinda killed my motivation a bit! 😅 (even though it's probably not as bad as I'm making it seem it is)

A part of me is hoping the protocol docs will get updated for the actual v5 release, but if not I suppose the best approach would be to slowly start adding manual structs for the more common requests & events until it is.

**Comment by andreykaipov** [comment]
Hello! I managed to get a working library out for v5! https://github.com/andreykaipov/goobs/releases

The protocol changed a lot so I imagine updating this CLI will take a little bit to find out what the corresponding methods are. For example I took a look at the `profile.go` and now instead of `client.Profiles.ListProfiles()`, it'd be `client.Config.GetProfileList()`.

If anybody takes a stab at it please do lemme know if y'all run into anything!
